### PR TITLE
hackweek: Relay dashboard

### PR DESCRIPTION
**Pull request by jjbayer**
### Add Web UI Dashboard to Relay. 

```[tasklist]
### TODOs
- [x] @jjbayer keep more log entries (ring buffer)
- [x] @olksdr css (toolbar, relay logo, ...)
- [x] @olksdr add colors to logs
- [x] @jjbayer visualization for statsd
- [ ] tools (`process-event`)
- [ ] tokio top
```

```[tasklist]
### Bugs
- [x] Frontend should keep trying to (re)connect to server
- [x] Backend must not run out of file descriptors
- [x] 100% CPU usage makes page unusable after a while
```

### How this looks like now

Currently we can see:
* all `statsd` metrics - the graphs will appear as soon as there are any data produced
* the logs will be also shown in the UI and automatically will be scrolled to the newest ones 

image

### CODE REVIEW

File: relay-admin-dashboard/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,19 @@
+[package]
+name = "relay-admin-dashboard"
```
Comment (olksdr) [inline relay-admin-dashboard/Cargo.toml]:
```suggestion
name = "relay-dashboard"
```

To be honest I would keep the name simpler, just `relay-dashboard` and omit all the `admin` from any of the places.

File: relay-admin-dashboard/README.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
+# Relay Admin Dashboard
+
+WIP
```
Comment (olksdr) [inline relay-admin-dashboard/README.md]: Before merging we might want to add at least some top level description here

File: relay-admin-dashboard/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+use yew::prelude::*;
+use yew_router::prelude::*;
+
+use crate::utils::buffering_socket;
+
+mod stats;
+mod utils;
+
+const RELAY_URL: &str = "localhost:3001"; // TODO: make configurable
```
Comment (olksdr) [inline relay-admin-dashboard/src/main.rs]: we should be able to take this from the browser window location

File: relay-log/src/setup.rs [hunk 4]
```diff
@@ -7,6 +7,9 @@ use serde::{Deserialize, Serialize};
 use tracing::{level_filters::LevelFilter, Level};
 use tracing_subscriber::{prelude::*, EnvFilter, Layer};
 
+#[cfg(feature = "admin")]
```
Comment (olksdr) [inline relay-log/src/setup.rs]: and also the feature flag I would rename to `dashboard`

File: relay-server/src/endpoints/stats.rs [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    let Some(rx) = relay_statsd::init_basic() else {
+        relay_log::error!("MetricsClient has already been set up");
+        return;
+    };
+    loop {
+        let rx = rx.clone();
+        if let Ok(Ok(bytes)) = spawn_blocking(move || rx.recv()).await {
+            let message = String::from_utf8_lossy(&bytes).to_string();
+            let res = socket.send(message.into()).await;
+            if res.is_err() {
```
Comment (olksdr) [inline relay-server/src/endpoints/stats.rs:21]: The error happens only on client disconnect or is there a specific error we should be looking for?

**Review by olksdr — APPROVED** [review]
Let's 🚢  it!

File: relay-admin-dashboard/Cargo.toml (hunk 1 above)
Comment (jjbayer) [reply relay-admin-dashboard/Cargo.toml]: Agreed!

**Review by iker-barriocanal — APPROVED** [review]

File: relay-server/src/endpoints/stats.rs (hunk 5 above)
Comment (jjbayer) [reply relay-server/src/endpoints/stats.rs:21]: I copied this straight from the [example](https://docs.rs/axum/latest/axum/extract/ws/index.html#example) so it Should Be Fine™.

**Pull request merged**